iib/web/api_v1.py

```diff
@@ -37,3 +38,83 @@ def add_bundles():
+
+
+@api_v1.route('/builds/<int:request_id>', methods=['PATCH'])
+@login_required
+def patch_request(request_id):
+    """
+    Modify the given request.
+
+    :param int request_id: the request ID from the URL
+    :return: a Flask JSON response
+    :rtype: flask.Response
+    :raise NotFound: if the request is not found
+    :raise ValidationError: if the JSON is invalid
+    """
+    allowed_users = flask.current_app.config['IIB_WORKER_USERNAMES']
+    # current_user.is_authenticated is only ever False when auth is disabled
+    if current_user.is_authenticated and current_user.username not in allowed_users:
+        raise Unauthorized('This API endpoint is restricted to IIB workers')
+
+    payload = flask.request.get_json()
+    if not isinstance(payload, dict):
+        raise ValidationError('The input data must be a JSON object')
+
+    if not payload:
+        raise ValidationError('At least one key must be specified to update the request')
+
+    valid_keys = {
+        'add_arches', 'binary_image_resolved', 'from_index_resolved', 'index_image',
+        'state', 'state_reason'
+    }
+    invalid_keys = payload.keys() - valid_keys
+    if invalid_keys:
+        raise ValidationError(
+            'The following keys are not allowed: {}'.format(', '.join(invalid_keys)))
+
+    for key, value in payload.items():
+        if key == 'add_arches':
+            Architecture.validate_architecture_json(value)
+
+        if key in ('binary_image_resolved', 'from_index_resolved', 'index_image'):
+            if not value or not isinstance(value, str):
+                raise ValidationError(f'The value for "{key}" must be a non-empty string')
+
+    if 'state' in payload and 'state_reason' not in payload:
+        raise ValidationError('The "state_reason" key is required when "state" is supplied')
+    elif 'state_reason' in payload and 'state' not in payload:
+        raise ValidationError('The "state" key is required when "state_reason" is supplied')
+
+    request = Request.query.get_or_404(request_id)
+    if 'state' in payload and 'state_reason' in payload:
+        new_state = payload['state']
+        new_state_reason = payload['state_reason']
+        # This is to protect against a Celery task getting executed twice and setting the
+        # state each time
+        if request.state.state == new_state and request.state.state_reason == new_state_reason:
+            flask.current_app.logger.info('Not adding a new state since it matches the last state')
+        else:
+            request.add_state(new_state, new_state_reason)
+
+    for key in ('binary_image_resolved', 'from_index_resolved', 'index_image'):
+        if key not in payload:
+            continue
+        key_value = payload.get(key, None)
+        key_object = Image.get_or_create(key_value)
+        if key_object not in getattr(request, key):
+            getattr(request, key).append(key_object)
+
+    if 'add_arches' in payload:
```

Comment:
This if statement can be removed since your for loop guards against the key not being set.